### PR TITLE
Backend for getting logs of a trial

### PR DESCRIPTION
**Pull request by d-gol**
Implementation of the backend for fetching logs, as a part of https://github.com/kubeflow/katib/issues/1764.  

Providing a route `/katib/fetch_trial_logs/` to obtain logs for a specific trial.  
Logs are obtained from a master pod.

### CODE REVIEW

File: pkg/new-ui/v1beta1/backend.go
Comment (johnugeorge) [inline pkg/new-ui/v1beta1/backend.go:475]: How about using Trial Spec Primary Pod Labels?

https://github.com/kubeflow/katib/blob/master/pkg/webhook/v1beta1/pod/inject_webhook.go#L129 

/cc @andreyvelich

Comment (d-gol) [reply pkg/new-ui/v1beta1/backend.go:475]: Agree, I can change it. Did you mean something like this?
```
selectionLabel := "training.kubeflow.org/job-name=" + trialName

for primaryKey, primaryValue := range trial.Spec.PrimaryPodLabels {
	selectionLabel = selectionLabel + "," + primaryKey + "=" + primaryValue
}
```

Comment (d-gol) [reply pkg/new-ui/v1beta1/backend.go]: Edited here: https://github.com/kubeflow/katib/pull/2039/commits/4d979d9199c0c6470aec5573c4611926b045510c

File: pkg/new-ui/v1beta1/backend.go
Comment (andreyvelich) [inline pkg/new-ui/v1beta1/backend.go:479]: I think you can remove `selectionLabel` variable.
Primary Pod Labels already has necessary labels to get the pods (e.g. [the default for Training Operator](https://github.com/kubeflow/katib/blob/master/pkg/apis/controller/experiments/v1beta1/constants.go#L41))
Or [the required for Argo Workflow example](https://github.com/kubeflow/katib/blob/master/examples/v1beta1/argo/argo-workflow.yaml#L35).

File: pkg/new-ui/v1beta1/backend.go
Comment (andreyvelich) [inline pkg/new-ui/v1beta1/backend.go:441]: Should you remove this ?
```suggestion
```

File: pkg/new-ui/v1beta1/backend.go
Comment (andreyvelich) [inline pkg/new-ui/v1beta1/backend.go:437]:
```suggestion
	log.Println("Requesting logs")
```

File: pkg/new-ui/v1beta1/backend.go
Comment (andreyvelich) [inline pkg/new-ui/v1beta1/backend.go:469]: Can we use [the Controller Runtime client from the `KatibUIHandler`](https://github.com/kubeflow/katib/blob/4d979d9199c0c6470aec5573c4611926b045510c/pkg/util/v1beta1/katibclient/katib_client.go#L60-L81) to get the Pod logs ?

File: pkg/new-ui/v1beta1/backend.go
Comment (andreyvelich) [inline pkg/new-ui/v1beta1/backend.go:477]: Should we verify the auth here to check if user can get the Pod logs, similar to [other API calls](https://github.com/kubeflow/katib/blob/master/pkg/new-ui/v1beta1/backend.go#L108) ?
cc @apo-ger @kimwnasptd

File: pkg/new-ui/v1beta1/backend.go
Comment (andreyvelich) [inline pkg/new-ui/v1beta1/backend.go:493]: Let's use the example with commit so it will never be changed ?
```suggestion
An example can be found here: https://github.com/kubeflow/katib/blob/7bf39225f7235ee4ba6cf285ecc2c455c6471234/examples/v1beta1/argo/argo-workflow.yaml#L33`
```

File: pkg/new-ui/v1beta1/backend.go
Comment (andreyvelich) [inline pkg/new-ui/v1beta1/backend.go:502]: You can get the constant from here: https://github.com/kubeflow/katib/blob/4d979d9199c0c6470aec5573c4611926b045510c/pkg/metricscollector/v1beta1/common/const.go#L53.

File: pkg/new-ui/v1beta1/backend.go
Comment (andreyvelich) [inline pkg/new-ui/v1beta1/backend.go:500]: I think, you can skip this check.
Trial instance has [Metrics Collector spec](https://github.com/kubeflow/katib/blob/4d979d9199c0c6470aec5573c4611926b045510c/pkg/apis/controller/trials/v1beta1/trial_types.go#L47).
You can just do this: 
```go
if trial.spec.metricsCollector.collector.kind == apis.StdOut {
 podLogOpts.Container = const. MetricLoggerCollectorContainerName
}
```

**Review by andreyvelich — COMMENTED** [review]
Thank you for implementing this @d-gol!
I left few comments.

File: pkg/new-ui/v1beta1/backend.go
Comment (apo-ger) [reply pkg/new-ui/v1beta1/backend.go:477]: @andreyvelich I think we should. This is the required authorization check before fetching the trial:

```go
user, err := IsAuthorized(consts.ActionTypeGet, namespace, consts.PluralTrial, "", trialName, trialv1beta1.SchemeGroupVersion, k.katibClient.GetClient(), r)
if user == "" && err != nil {
	log.Printf("No user provided in kubeflow-userid header.")
	http.Error(w, err.Error(), http.StatusUnauthorized)
	return
} else if err != nil {
	log.Printf("The user: %s is not authorized to delete tial: %s in namespace: %s \n", user, trialName, namespace)
	http.Error(w, err.Error(), http.StatusForbidden)
	return
}

```
Similar to what we do in [new-ui/v1beta1/hp.go](https://github.com/kubeflow/katib/blob/bd91301d36051d4cf04fe88bb3751591df4141b9/pkg/new-ui/v1beta1/hp.go#L98-L107)

File: pkg/new-ui/v1beta1/backend.go
Comment (d-gol) [reply pkg/new-ui/v1beta1/backend.go:479]: Don't we also need a selection based on the name of the trial? Without the `training.kubeflow.org/job-name` label, we would get all the master replica pods in the namespace.

File: pkg/new-ui/v1beta1/backend.go
Comment (d-gol) [reply pkg/new-ui/v1beta1/backend.go:469]: From my understanding based on this issue https://github.com/kubernetes-sigs/controller-runtime/issues/452, it is not possible to obtain logs using the controller runtime client, we need a Kubernetes client. If I'm missing something, please let me know.

**Comment by tenzen-y** [comment]
@d-gol Thanks for your effort.
We merged #2047 with the master branch to mutate the trial name label to pods. So, can you rebase this PR?

File: pkg/new-ui/v1beta1/backend.go
Comment (d-gol) [reply pkg/new-ui/v1beta1/backend.go:630]: Thanks @apo-ger for the code snippet!

File: pkg/new-ui/v1beta1/backend.go
Comment (d-gol) [reply pkg/new-ui/v1beta1/backend.go]: Edited the label https://github.com/kubeflow/katib/pull/2047 was merged.

**Comment by d-gol** [comment]
Thank you @tenzen-y and @andreyvelich, rebased it now.

File: pkg/new-ui/v1beta1/backend.go
Comment (d-gol) [reply pkg/new-ui/v1beta1/backend.go]: Thanks @tenzen-y, added the comment.

**Comment by johnugeorge** [comment]
Thanks Dejan

/lgtm

**Review by tenzen-y — COMMENTED** [review]
@d-gol Thanks for driving this!

/lgtm
/assign @andreyvelich

File: pkg/new-ui/v1beta1/backend.go
Comment (d-gol) [reply pkg/new-ui/v1beta1/backend.go:597]: Yes, I agree. Added authorization checks for listing the pods and getting the logs. Had to reorganize the code a bit to fit in the additional checks, but the logic is the same. Adding as a separate commit, in the end we can squash the commits.

File: pkg/new-ui/v1beta1/backend.go
Comment (andreyvelich) [inline pkg/new-ui/v1beta1/backend.go:608]: Should we use group version for pods similar to ConfigMaps (https://github.com/kubeflow/katib/blob/365154b2ebfcdc5fdacdbd60c4dc813fc9c60c39/pkg/new-ui/v1beta1/backend.go#L359)  ?
`corev1.SchemeGroupVersion` ?

File: manifests/v1beta1/components/ui/rbac.yaml
Comment (andreyvelich) [inline manifests/v1beta1/components/ui/rbac.yaml:33]: I think, we should also update the `kubeflow-user-edit` role: https://github.com/kubeflow/katib/blob/master/manifests/v1beta1/installs/katib-with-kubeflow/kubeflow-katib-roles.yaml#L22-L36 with this permission.

cc @kimwnasptd

File: pkg/new-ui/v1beta1/backend.go
Comment (andreyvelich) [inline pkg/new-ui/v1beta1/backend.go:641]: @kimwnasptd @apo-ger Please review this change. Does it sound good for you ?

File: pkg/new-ui/v1beta1/backend.go
Comment (d-gol) [reply pkg/new-ui/v1beta1/backend.go]: Thank you, edited!

File: manifests/v1beta1/components/ui/rbac.yaml
Comment (d-gol) [reply manifests/v1beta1/components/ui/rbac.yaml:33]: Edited this as well.

File: pkg/new-ui/v1beta1/backend.go
Comment (d-gol) [reply pkg/new-ui/v1beta1/backend.go]: Let's wait for the review on this one.

Comment (apo-ger) [reply pkg/new-ui/v1beta1/backend.go]: Hey @andreyvelich @d-gol! The changes look good.

Comment (d-gol) [reply pkg/new-ui/v1beta1/backend.go]: Thanks @apo-ger!

File: pkg/new-ui/v1beta1/backend.go
Comment (apo-ger) [inline pkg/new-ui/v1beta1/backend.go:593]: @d-gol Just a minor comment here. Maybe add checks here similar to what we do [here](https://github.com/d-gol/katib/blob/4afe97603ef2ffbe868c42371cf6716e431754bd/pkg/new-ui/v1beta1/hp.go#L196-L213)

File: pkg/new-ui/v1beta1/backend.go
Comment (kimwnasptd) [inline pkg/new-ui/v1beta1/backend.go:611]: Slightly orthogonal to the PR, but I think the function signature `user, err = IsAuthorized(...)` is not ideal. We end up getting the same information and do duplicate checks on errors depending on the user value.

Why not have a distinct function for getting the current user and do this check once? Then `IsAuthorized(user, ...)` will only be responsible for the SubjectAccessReviews check.

Or at least for this PR, we could only check if the returned user is not "" only once, the first time we call this function

File: pkg/new-ui/v1beta1/backend.go
Comment (kimwnasptd) [inline pkg/new-ui/v1beta1/backend.go:631]: I'd be OK even if we'd remove the check for whether the user can list Pods. IMO just checking if the User has permissions to:
1. GET the specific Trial
2. GET the logs of each specific underlying Pod

would be enough. Even in plain K8s a user could have permissions to GET the logs of a Pod, but without having permissions to list all the Pods in the namespace

File: pkg/new-ui/v1beta1/backend.go
Comment (kimwnasptd) [inline pkg/new-ui/v1beta1/backend.go:688]: I understand that the backend will end up returning a string value, that is the logs of the "master" Pod of the Trial?

Question 1: Is there always a master pod or do we expect this to fail for some cases?

Comment (kimwnasptd) [inline pkg/new-ui/v1beta1/backend.go:688]: Question 2: Would we expect in the future to return logs from other worker pods?

If that's the case I'd propose that the backend actually returns a JSON type response like
```
logs: {
    master: "..."
}
```

**Review by kimwnasptd — COMMENTED** [review]
@d-gol @andreyvelich apologies for the late reply. Mostly have some questions around the return type and the structure of the response data.

File: pkg/new-ui/v1beta1/backend.go
Comment (d-gol) [reply pkg/new-ui/v1beta1/backend.go:611]: Agree, we are checking twice the user. I think we can proceed fixing the auth after we merge this PR? Then we can have a separate PR to improve the authentication in the entire file. Or we can do the other way around?

File: pkg/new-ui/v1beta1/backend.go
Comment (d-gol) [reply pkg/new-ui/v1beta1/backend.go:631]: Thanks @kimwnasptd , I can change this.

File: pkg/new-ui/v1beta1/backend.go
Comment (d-gol) [reply pkg/new-ui/v1beta1/backend.go:688]: I think the consensus is there will be a master pod. Please correct me if I'm wrong @johnugeorge @andreyvelich

Comment (d-gol) [reply pkg/new-ui/v1beta1/backend.go:688]: This is a good idea, if we want to change in the future. @johnugeorge @andreyvelich what do you think?

**Comment by tenzen-y** [comment]
@d-gol Can you rebase this since we have merged #2064 to fix CI into the master branch?

File: pkg/new-ui/v1beta1/backend.go
Comment (andreyvelich) [reply pkg/new-ui/v1beta1/backend.go:688]: @kimwnasptd We are using [PrimaryPodLabel to identify the pod for getting the logs](https://github.com/kubeflow/katib/pull/2039/files#diff-368f67643a9976088bc884f3f659f60407978206a4142bd75b0902411f2d4e72R707)

Comment (andreyvelich) [reply pkg/new-ui/v1beta1/backend.go:688]: I like @kimwnasptd idea, let's add Primary Pod Label to the JSON response.

Comment (d-gol) [reply pkg/new-ui/v1beta1/backend.go:688]: Hey @andreyvelich did you mean the result to be in the form: 
``` 
logs: {
    master: "..."
}
```
or something else?
We can have multiple primary pod labels, did you mean to also add them as key value pairs?

Comment (johnugeorge) [reply pkg/new-ui/v1beta1/backend.go:688]: master is bit confusing term. Eg: There can be job with just workers where worker0 acts as the master. If we really need to add pod info, pod name might be better.

Comment (andreyvelich) [reply pkg/new-ui/v1beta1/backend.go:688]: @d-gol It's not always "master" label for the pod that we get labels.
For example, for Argo we get pod with [`katib.kubeflow.org/model-training: true` label](https://github.com/kubeflow/katib/blob/master/examples/v1beta1/argo/argo-workflow.yaml#L35).
Maybe pod name to include in the response make sense as @johnugeorge mentioned.

Comment (andreyvelich) [reply pkg/new-ui/v1beta1/backend.go:688]: I am ok with creating separate issue to track improvements for this API response (e.g. add trial name to the response).
So we can merge this PR and unblock UI team to start working on the UI changes to have this feature in the next release.
What do you think @d-gol @kimwnasptd @johnugeorge ?

Comment (johnugeorge) [reply pkg/new-ui/v1beta1/backend.go:688]: I agree. +1 to merge

Comment (d-gol) [reply pkg/new-ui/v1beta1/backend.go:688]: I agree to merge it, and later we can improve the API response with more information if needed. So again, to clarify, we want to merge this PR with a simple string response (current implementation)? Or in the form of json, like below?
```
{
    "pod_name": logs
}
```

**Comment by johnugeorge** [comment]
@d-gol We can merge this. Can you create an issue to track the Json response discussion ? Also, please do a rebase

**Comment by d-gol** [comment]
@johnugeorge great, I will rebase it. Also, I can remove the check for listing pods, as Kimonas suggested - https://github.com/kubeflow/katib/pull/2039#discussion_r1048278596

**Comment by d-gol** [comment]
Rebased @johnugeorge  @andreyvelich should I squash the commits, or will you do it on merge? There were many commits during the discussions, but I think they can all be squashed.

**Comment by andreyvelich** [comment]
> There were many commits during the discussions, but I think they can all be squashed.

@d-gol They will be automatically squashed by tide before the merge.
Thank you for the updates!
/lgtm

/assign @johnugeorge @tenzen-y @kimwnasptd

**Comment by d-gol** [comment]
@andreyvelich great, thank you!

**Comment by johnugeorge** [comment]
@d-gol Can you try a failing e2e test locally?

**Comment by d-gol** [comment]
@johnugeorge sure, checking it.

File: pkg/new-ui/v1beta1/backend.go
Comment (tenzen-y) [inline pkg/new-ui/v1beta1/backend.go:623]: nit: 
Do you have any reason for using `context.TODO()`?
It might be better to use `context.Background()` or to take context out of `r *http.Request` (r.Context()).

**Review by tenzen-y — COMMENTED** [review]
@d-gol Thanks for implementing this powerful feature!

LGTM
Although I wonder why our E2E failed.

File: pkg/new-ui/v1beta1/backend.go
Comment (d-gol) [reply pkg/new-ui/v1beta1/backend.go:623]: Thanks a lot @tenzen-y, and thank you for all the help!

Actually no particular reason, I will change now to `context.Background()`.

For E2E, strange, they were passing on each previous commit, this one today should not affect them.
Checking locally as well.

**Comment by tenzen-y** [comment]
@johnugeorge @andreyvelich @d-gol Same errors seem to occur for `E2E Test with Katib UI, random search, and postgres / e2e` and `E2E Test with mxnet-mnist / e2e` in #2060 and #2067.

**Comment by tenzen-y** [comment]
It seems that errors are caused by mxnet-mnist image. You can reproduce by `ytenzen/mxnet-mnist:debug-error`.

**Comment by tenzen-y** [comment]
ASAP, I will create a PR to fix this issue.

**Comment by tenzen-y** [comment]
Blocked by: #2070

**Comment by tenzen-y** [comment]
@d-gol Can you rebase since we fixed CI?

**Comment by d-gol** [comment]
@tenzen-y done, thank you for fixing the CI!

**Review by tenzen-y — APPROVED** [review]
@d-gol Thank you for the update!
/lgtm
/approve

**Comment by d-gol** [comment]
@tenzen-y great, thanks a lot! 
And thanks everyone for all the help with this @andreyvelich @johnugeorge @kimwnasptd @apo-ger

**Pull request merged**
